c/src/ml-api-service-db.cc

```diff
@@ -0,0 +1,268 @@
+  virtual void getPipelineDescription (std::string name,
+      std::string & pipeline_description) override;
+  virtual void delPipelineDescription (std::string name);
+
+private:
+  std::string path;
+  leveldb_t *db_obj;
+  leveldb_readoptions_t *db_roptions;
+  leveldb_writeoptions_t *db_woptions;
+};
+
+static IMLServiceDB *instance = nullptr;
+
+/**
+ * @brief Get the instance object of MLServiceLevelDB
+ * @return IMLServiceDB* instance of MLServiceLevelDB
+ */
+static IMLServiceDB *
+get_instance ()
+{
+  if (!instance) {
+    instance = new MLServiceLevelDB (ML_DATABASE_PATH);
+    instance->connectDB ();
+  }
+
+  return instance;
+}
+
+/**
+ * @brief Construct a new MLServiceLevelDB object
+ * @param path database path
+ */
+MLServiceLevelDB::MLServiceLevelDB (std::string path)
+:  path (path)
+{
+  db_obj = nullptr;
+}
+
+/**
+ * @brief Destroy the MLServiceLevelDB object
+ */
+MLServiceLevelDB::~MLServiceLevelDB ()
+{
+}
+
+/**
+ * @brief Connect the level DB and initialize the private variables.
+ */
+void
+MLServiceLevelDB::connectDB ()
+{
+  char *err = nullptr;
+  leveldb_options_t *db_options;
+
+  db_options = leveldb_options_create ();
+  leveldb_options_set_create_if_missing (db_options, 1);
+
+  db_obj = leveldb_open (db_options, path.c_str (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Error! Failed to open Database : %s", err);
+    leveldb_free (err);
+    throw std::runtime_error ("Failed to connectDB()!");
+  }
+
+  db_roptions = leveldb_readoptions_create ();
+  db_woptions = leveldb_writeoptions_create ();
+  leveldb_writeoptions_set_sync (db_woptions, 1);
+
+  return;
+}
+
+/**
+ * @brief Set the pipeline description with the given name.
+ * @note If the name already exists, the pipeline description is overwritten.
+ * @param[in] name Unique name to retrieve the associated pipeline description.
+ * @param[in] pipeline_description The pipeline description to be stored.
+ */
+void
+MLServiceLevelDB::setPipelineDescription (std::string name,
+    std::string pipeline_description)
+{
+  char *err = NULL;
+
+  leveldb_put (db_obj, db_woptions, name.c_str (), name.size (),
+      pipeline_description.c_str (), pipeline_description.size (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Failed to put %s : %s", name.c_str (), err);
+    leveldb_free (err);
+    throw std::runtime_error ("Failed to setPipelineDescription()!");
+  }
+
+  return;
+}
+
+/**
+ * @brief Get the pipeline description with the given name.
+ * @param[in] name The unique name to retrieve.
+ * @param[out] pipeline_description The pipeline corresponding with the given name.
+ */
+void
+MLServiceLevelDB::getPipelineDescription (std::string name,
+    std::string & pipeline_description)
+{
+  char *err = NULL;
+  char *value = NULL;
+  gsize read_len;
+
+  value = leveldb_get (db_obj, db_roptions, name.c_str (), name.size (),
+      &read_len, &err);
+  if (!value) {
+    _ml_loge ("Fail to find the key %s", name.c_str ());
+    leveldb_free (err);
```

Comment:
You didn't check for err here. You've checked ```!value``` only. If err is available, use it for error message, too!

Reply:
Handle the case of (!value && err). (print all available info for app writers!)

Reply:
If the key is not found in the database, `leveldb_get()` returns `NULL` without an error message.
However, the err is set in case that a database operation error such as file corruption or permission.
Anyway, I will update the error message and clean up this code.